mypy/test/data.py

```diff
@@ -470,14 +471,29 @@ def pytest_pycollect_makeitem(collector: Any, name: str, obj: Any) -> Any:
     return MypyDataSuite(name, parent=collector)
 
 
+# An instance of this class is created for each subclass of DataSuite encountered by pytest;
+# pytest.Class machinery sets instance attribute obj: Type[DataSuite] to that subclass.
+# Beyond this, we should not rely on obj-related machinery, since it's undocumented and implicit.
 class MypyDataSuite(pytest.Class):  # type: ignore  # inheriting from Any
+    # Will yield collectors that represent individual case files.
+    def collect(self) -> Iterator['MypyDataCaseFile']:
+        for case_file in self.obj.case_files:
+            yield MypyDataCaseFile(case_file, self)
+
+
+# An instance of this class is tied (via .name and .parent) to a specific case file.
+class MypyDataCaseFile(pytest.Collector):  # type: ignore  # inheriting from Any
     def collect(self) -> Iterator['MypyDataCase']:
-        for case in self.obj.cases():
+        for case in parse_test_cases(os.path.join(test_data_prefix, self.name),
+                                     None, test_temp_dir, True):
             yield MypyDataCase(case.name, self, case)
 
 
+# An instance of this class is tied (via .obj) to a specific case.
+# Note that despite the use of the obj attribute, the obj-related pytest machinery is not invoked.
+# TODO: rename obj into case.
```

Comment:
TLDR; In the latest commit in this PR (which I'm  about to push), I'm no longer relying on this mechanism at all.

Below I'll describe how it works on the current master (and in my own earlier commit in this PR):

By "obj-related pytest machinery" I mean the a mechanism that populates `.obj` attribute in each instance of `pytest.Class` with the corresponding class object from which this item was created. For example, when pytest encounters `class TypeCheckSuite`, it calls our hook `pytest_pycollect_makeitem` to yield tests cases from it. The hook yields instances of `MypyDataSuite`, and [those instances magically end up with an `.obj` attribute](https://github.com/python/mypy/blob/e05f6f3db9597cced43ae71997cd6ae440311bd2/mypy/test/data.py#L500) - even though we never explicitly assigned anything to `obj`. That attribute is populated because `MypyDataSuite` indirectly derives from `pytest.PyobjMixin`, which [implements `obj` property](https://github.com/pytest-dev/pytest/blob/f0541b685bb7d27ea87eb8fd8bddbd0280a9b48d/_pytest/python.py#L200L212) that behaves as described.

AFAICT, the existence of `obj` is undocumented, so it may have been intended for internal pytest use only.

The additional confusion arose because in another class (`MypyDataCase`), we also use an `obj` attribute, which is populate explicitly through a constructor.